native-lib-loader 2.2.0 -> 2.3.0 Regression

Hi. I have a project that uses nar-maven-plugin's native-lib-loader capabilities.

Everything works fine with the latest nar-maven-plugin (3.5.2) and native-lib-loader 2.2.0. However, when I tried to upgrade native-lib-loader dependency to version 2.3.0, `mvn clean install` fails with:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project neanderthal-mkl: Compilation failure
[ERROR] /home/dragan/workspace/clojure/neanderthal-native/neanderthal-mkl/target/nar/nar-generated/uncomplicate/neanderthal/internal/host/NarSystem.java:[46,44] constructor DefaultJniExtractor in class org.scijava.nativelib.DefaultJniExtractor cannot be applied to given types;
[ERROR]   required: java.lang.Class<?>
[ERROR]   found: java.lang.Class<uncomplicate.neanderthal.internal.host.NarSystem>,java.lang.String
[ERROR]   reason: actual and formal argument lists differ in length
[ERROR] 
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project neanderthal-mkl: Compilation failure
/home/dragan/workspace/clojure/neanderthal-native/neanderthal-mkl/target/nar/nar-generated/uncomplicate/neanderthal/internal/host/NarSystem.java:[46,44] constructor DefaultJniExtractor in class org.scijava.nativelib.DefaultJniExtractor cannot be applied to given types;
  required: java.lang.Class<?>
  found: java.lang.Class<uncomplicate.neanderthal.internal.host.NarSystem>,java.lang.String
  reason: actual and formal argument lists differ in length

    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
/home/dragan/workspace/clojure/neanderthal-native/neanderthal-mkl/target/nar/nar-generated/uncomplicate/neanderthal/internal/host/NarSystem.java:[46,44] constructor DefaultJniExtractor in class org.scijava.nativelib.DefaultJniExtractor cannot be applied to given types;
  required: java.lang.Class<?>
  found: java.lang.Class<uncomplicate.neanderthal.internal.host.NarSystem>,java.lang.String
  reason: actual and formal argument lists differ in length

    at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute (AbstractCompilerMojo.java:1161)
    at org.apache.maven.plugin.compiler.CompilerMojo.execute (CompilerMojo.java:168)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

--- a/pom.xml
+++ b/pom.xml
@@ -538,6 +538,7 @@
                 <setupInclude>it0010-lib-static/pom.xml</setupInclude>
                 <setupInclude>it0020-lib-3rdparty/pom.xml</setupInclude>
                 <setupInclude>it0026-native-lib-loader/pom.xml</setupInclude>
+                <setupInclude>it0026-native-lib-loader-2.3.0/pom.xml</setupInclude>
                 <setupInclude>it0031-push-deps-to-lowest-order-agg/pom.xml</setupInclude>
                 <setupInclude>it0034-upstream-aol1-multi-aol/pom.xml</setupInclude>
                 <setupInclude>it0034-upstream-aol2-multi-aol/pom.xml</setupInclude>

--- a/src/it/it0026-native-lib-loader-2.3.0/pom.xml
+++ b/src/it/it0026-native-lib-loader-2.3.0/pom.xml
@@ -1,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  #%L
+  Native ARchive plugin for Maven
+  %%
+  Copyright (C) 2002 - 2014 NAR Maven Plugin developers.
+  %%
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  #L%
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.github.maven-nar.its.nar</groupId>
+        <artifactId>it-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../it-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>it0026-native-lib-loader-2.3.0</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>nar</packaging>
+
+    <name>NAR with native-lib-loader</name>
+    <description>
+        NAR with native-lib-loader
+    </description>
+
+    <build>
+        <defaultGoal>integration-test</defaultGoal>
+        <plugins>
+            <plugin>
+                <groupId>@project.groupId@</groupId>
+                <artifactId>nar-maven-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <c>
+                        <debug>true</debug>
+                    </c>
+                    <libraries>
+                        <library>
+                            <type>jni</type>
+                            <narSystemPackage>it0026.test</narSystemPackage>
+                            <linkCPP>false</linkCPP>
+                        </library>
+                    </libraries>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.scijava</groupId>
+            <artifactId>native-lib-loader</artifactId>
+            <version>2.3.0</version>
+        </dependency>
+        <!-- for native-lib-loader -->
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>1.7.6</version>
+        </dependency>
+    </dependencies>
+</project>

--- a/src/it/it0026-native-lib-loader-2.3.0/src/main/c/hello.c
+++ b/src/it/it0026-native-lib-loader-2.3.0/src/main/c/hello.c
@@ -1,0 +1,27 @@
+/*
+ * #%L
+ * Native ARchive plugin for Maven
+ * %%
+ * Copyright (C) 2002 - 2014 NAR Maven Plugin developers.
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include <stdio.h>
+#include "it0026_test_NarSystem.h"
+
+JNIEXPORT jbyte JNICALL Java_it0026_test_Hello_test
+  (JNIEnv *env, jclass clazz)
+{
+	return (jbyte)108;
+}

--- a/src/it/it0026-native-lib-loader-2.3.0/src/main/java/it0026/test/Hello.java
+++ b/src/it/it0026-native-lib-loader-2.3.0/src/main/java/it0026/test/Hello.java
@@ -1,0 +1,33 @@
+/*
+ * #%L
+ * Native ARchive plugin for Maven
+ * %%
+ * Copyright (C) 2002 - 2014 NAR Maven Plugin developers.
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package it0026.test;
+
+public class Hello {
+	static
+	{
+		NarSystem.loadLibrary();
+	}
+
+	public native static byte test();
+
+	public static void main(final String... args) {
+		System.exit((int) test());
+	}
+}

--- a/src/it/it0026-native-lib-loader-2.3.0/src/test/java/it0026/test/HelloTest.java
+++ b/src/it/it0026-native-lib-loader-2.3.0/src/test/java/it0026/test/HelloTest.java
@@ -1,0 +1,32 @@
+/*
+ * #%L
+ * Native ARchive plugin for Maven
+ * %%
+ * Copyright (C) 2002 - 2014 NAR Maven Plugin developers.
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package it0026.test;
+
+import it0026.test.Hello;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class HelloTest
+{
+	@Test public final void testNativeMethod()
+	{
+		Assert.assertEquals(108, Hello.test());
+	}
+}

--- a/src/it/it0026-native-lib-loader-2.3.0/verify.bsh
+++ b/src/it/it0026-native-lib-loader-2.3.0/verify.bsh
@@ -1,0 +1,75 @@
+/*
+ * #%L
+ * Native ARchive plugin for Maven
+ * %%
+ * Copyright (C) 2002 - 2014 NAR Maven Plugin developers.
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+/*
+ * This script verifies that the native-lib-loader method to unpack and
+ * load native libraries works even with fully packaged .jar files.
+ */
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.InputStreamReader;
+
+nativeLibLoaderVersion = "2.3.0";
+sl4jApiVersion = "1.7.6";
+classPath = localRepositoryPath + "/org/scijava/native-lib-loader/"
+	+ nativeLibLoaderVersion + "/native-lib-loader-"
+	+ nativeLibLoaderVersion + ".jar"
+	+ File.pathSeparator
+  + localRepositoryPath + "/org/slf4j/slf4j-api/"
+  + sl4jApiVersion + "/slf4j-api-"
+  + sl4jApiVersion + ".jar";
+
+new File(basedir + "/target/").list(new FilenameFilter() {
+	public boolean accept(File dir, String name) {
+		if (name.startsWith("it0026-native-lib-loader-2.3.0-1.0-SNAPSHOT")) {
+			classPath += File.pathSeparator
+				+ new File(dir, name).getPath();
+		}
+		return false;
+	}
+});
+
+printInputStream(in) {
+	buffered = new BufferedReader(new InputStreamReader(in));
+	new Thread() {
+		public void run() {
+			for (;;) {
+				line = buffered.readLine();
+				if (line == null) break;
+				print(line);
+			}
+			buffered.close();
+		}
+	}.start();
+}
+
+print("Starting java -classpath '" + classPath + "' it0026.test.Hello");
+process = new ProcessBuilder(new String[] {
+		"java", "-classpath", classPath, "it0026.test.Hello"
+	}).start();
+printInputStream(process.getErrorStream());
+printInputStream(process.getInputStream());
+exitCode = process.waitFor();
+
+if (exitCode != 108) {
+	print("Unexpected exit code: " + exitCode);
+	return false;
+}

--- a/src/main/java/com/github/maven_nar/NarSystemMojo.java
+++ b/src/main/java/com/github/maven_nar/NarSystemMojo.java
@@ -7,9 +7,9 @@
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  * http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -139,16 +139,21 @@ public class NarSystemMojo extends AbstractNarMojo {
   }
 
   private boolean hasNativeLibLoaderAsDependency() {
+    return getNativeLibLoaderVersion() != null;
+  }
+
+
+  private String getNativeLibLoaderVersion() {
     for (MavenProject project = getMavenProject(); project != null; project = project.getParent()) {
       final List<Dependency> dependencies = project.getDependencies();
       for (final Dependency dependency : dependencies) {
         final String artifactId = dependency.getArtifactId();
         if ("native-lib-loader".equals(artifactId)) {
-          return true;
+          return dependency.getVersion();
         }
       }
     }
-    return false;
+    return null;
   }
 
   @Override
@@ -214,7 +219,7 @@ public class NarSystemMojo extends AbstractNarMojo {
           + "            System.load(unpacked.getPath());\n"
           + "        } else try {\n"
           + "            final String libPath = getLibPath(loader, aols, mappedNames);\n"
-          + "            final JniExtractor extractor = new DefaultJniExtractor(NarSystem.class, System.getProperty(\"java.io.tmpdir\"));\n"
+          + "            final JniExtractor extractor = "+ getJniExtractorCreationStatement() +"\n"
           + "            final File extracted = extractor.extractJni(libPath, fileName);\n"
           + "            System.load(extracted.getAbsolutePath());\n" + "        } catch (final Exception e) {\n"
           + "            e.printStackTrace();\n" + "            throw e instanceof RuntimeException ?\n"
@@ -276,5 +281,18 @@ public class NarSystemMojo extends AbstractNarMojo {
     if (this.buildContext != null) {
       this.buildContext.refresh(narSystem);
     }
+  }
+
+  private String getJniExtractorCreationStatement() {
+    String nativeLoaderVersion = getNativeLibLoaderVersion();
+    if (nativeLoaderVersion == null) {
+      throw new AssertionError("getJniExtractorCreationStatement() called, but nativel-lib-loader plugin is absent, or its version is unknown");
+    }
+    // versions 1.x.x are unavailable on Maven Central and Github, not checking against them
+    // new format was introduced in "2.3.0"
+    boolean oldFormat = nativeLoaderVersion.matches("2\\.[012]\\..*");
+    return oldFormat
+        ? "new DefaultJniExtractor(NarSystem.class, System.getProperty(\"java.io.tmpdir\"));"
+        : "new DefaultJniExtractor(NarSystem.class);";
   }
 }

